### PR TITLE
Fix processing of very large entries in queue

### LINKED ISSUE #16307: ZooKeeper coordination may fail permanently after multiple huge ALTERs are executed.
**Describe the bug:**

Create a table with very long column names or type names (type names can be long for legitimate reason, e.g. Enum).
Make table structure to be less than 1 MB but larger than 100 KB.

Create two replicas, turn one replica off. Execute at least 10 ALTER queries to slightly modify columns in the table.
Turn second replica on.

It will fail with:
```
<Debug> sandbox.taskdiskusage (ReplicatedMergeTreeQueue): Pulling 42 entries to queue: log-0001412404 - log-0001412445
...
<Error> void Coordination::ZooKeeper::receiveThread(): Code: 33, e.displayText() = DB::Exception: Cannot read all data. Bytes read: 0...
```
(there is no subsequent message `Pulled 42 entries to queue`)

Because it tries to make a transaction to move several `log` entries to `queue` (the code is in `ReplicatedMergeTreeQueue::pullLogsToQueue` function) and this transaction is larger than 1 MB.

After this message all other ZooKeeper queries will fail with `Session expired` error.
After the session to ZooKeeper is reinitialized, everything will repeat.

**Workaround:**

1 MB is the default limit for packet size in ZooKeeper. You should increase it with the parameter `-Djute.maxbuffer` in ZooKeeper.

**Notes:**

There was no such issue with ALTER columns queries before version 20.3. In version 20.3 ALTER of columns goes through the replication log, the change was added in #8701. It means that the issue may manifestate itself after upgrade to the versions 20.3+. But the root cause of the bug itself is present in every version, you can get similar effect with ALTER UPDATE or DELETE if the expression is very long (e.g. long list of constants).

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix processing of very large entries in replication queue. Very large entries may appear in ALTER queries if table structure is extremely large (near 1 MB). This fixes #16307.

Detailed description / Documentation draft:
Replication will make a progress before "Session expired" event and the issue will eventually fix automatically.
I struggle to write a test for this case.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Testflows - bad tests, @vzakaznikov

**Review by alesapin — APPROVED** [review]
Now LGTM

File: src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp [hunk 1]
```diff
@@ -490,14 +490,17 @@ int32_t ReplicatedMergeTreeQueue::pullLogsToQueue(zkutil::ZooKeeperPtr zookeeper
...
-        for (size_t entry_idx = 0, num_entries = log_entries.size(); entry_idx < num_entries; entry_idx += current_multi_batch_size)
+        for (size_t entry_idx = 0, num_entries = log_entries.size(); entry_idx < num_entries;)
         {
             auto begin = log_entries.begin() + entry_idx;
             auto end = entry_idx + current_multi_batch_size >= log_entries.size()
                 ? log_entries.end()
                 : (begin + current_multi_batch_size);
             auto last = end - 1;
 
+            /// Increment entry_idx before batch size increase (we copied at most current_multi_batch_size entries)
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp:501]: :+1:

**Comment by alexey-milovidov** [comment]
> test_multiple_disks

\- bad test, @excitoon

**Comment by alexey-milovidov** [comment]
> Yandex third-party checks (only for Yandex employees) 

Already broken.

**Pull request merged**